### ISSUE TITLE
📈 Plotting and Visualization Enhancements

### ISSUE THREAD

**Issue by scottviteri** [scottviteri/AttentionGuidedRL#30]
## 📋 Overview

This issue documents the **comprehensive plotting and visualization system** that provides rich insights into training dynamics, model behavior, and performance metrics through interactive plots and automated visualization generation.

**Related Commits:**
- Primary: `169e34b` - feat(plotting): enhanced debugging plots and documentation updates
- Demo: `93c3396` - feat: add demo_pipeline.py for data pipeline visualization
- Metrics: `71e2e75` - feat(main): add comprehensive debugging metrics and enhanced training analysis

### **1. Core Plotting Script: `generate_plots.py`**
Comprehensive standalone plotting tool with 20+ metric visualizations:

```bash
# Generate all plots for a training run
python generate_plots.py --input-dir runs/experiment_name --output-dir plots/

# Generate specific plot types
python generate_plots.py --input-dir runs/exp1 --metrics rewards,losses --output-dir plots/
```

### **2. Data Pipeline Visualization: `demo_pipeline.py`**
Educational tool demonstrating the functional data processing pipeline:

```bash
# Show data pipeline flow
python demo_pipeline.py --dataset wikipedia --batch-size 2 --steps 5

# Debug Twenty Questions pipeline
python demo_pipeline.py --dataset twenty_questions --verbose
```

### **3. Real-time Training Plots**
Integrated plotting during training with configurable intervals:

```bash
# Enable real-time plotting every 15 episodes
python src/main.py --enable-plotting --plot-interval 15
```

### **Training Progress Metrics**

| **Plot Type** | **Description** | **Key Insights** |
|---------------|-----------------|------------------|
| **Episode Rewards** | Mean, std, min, max rewards over time | Training progress, convergence |
| **Loss Components** | Policy loss, KL loss, total loss trends | Loss optimization dynamics |
| **Advantage Distribution** | Mean, std, percentile evolution | Policy learning effectiveness |
| **Clipping Ratios** | PPO clipping percentage over episodes | Training stability indicator |

### **Model Analysis Plots**

| **Plot Type** | **Description** | **Key Insights** |
|---------------|-----------------|------------------|
| **LoRA Gradients** | Gradient norms per layer over time | Training signal flow |
| **Similarity Scores** | Key-value similarity statistics | Selection diversity |
| **Learning Curves** | Training metrics with smoothing | Overall training health |
| **Performance Profiling** | Throughput and efficiency metrics | System optimization |

### **Demo Pipeline Flow**
The `demo_pipeline.py` script provides step-by-step visualization of data processing:

```python
# Sample output from demo_pipeline.py
Pipeline Demo: Wikipedia Dataset (batch_size=2, steps=5)

Step 1: Loading Wikipedia articles
├── Streaming from HuggingFace dataset
├── Article length filter: min 2000 tokens
└── Status: Found 15 suitable articles

Step 2: Tokenization and batching  
├── Batch 1: Articles 0-1, total tokens: 4,287
├── Batch 2: Articles 2-3, total tokens: 3,956
└── Complete batches only: 2 batches generated

Step 3: Key-Value extraction
├── Keys per batch: 10, Values per batch: 10
├── Key tokens shape: [2, 20], Value tokens: [2, 20]
└── Total KV pairs extracted: 20

Step 4: Embedding generation
├── Key embedding batch size: 4
├── Embedding dimension: 768
└── Total embeddings: [2, 10, 768]

Step 5: Stream composition
├── Pipeline: articles → filter → batch → extract → embed
├── Total processing time: 2.34s
└── Throughput: 8.5 KV pairs/second
```

### **Basic Plot Generation**
```bash
# Generate all available plots
python generate_plots.py --input-dir runs/my_experiment

# Generate specific metrics with custom styling
python generate_plots.py \
    --input-dir runs/exp1 \
    --metrics rewards,losses \
    --format svg \
    --theme dark
```

### **Pipeline Debugging**
```bash
# Debug Wikipedia pipeline
python demo_pipeline.py --dataset wikipedia --steps 3 --verbose

# Visualize Twenty Questions processing
python demo_pipeline.py --dataset twenty_questions --save-plots
```

### **Training with Visualization**
```bash
# Training with automatic plot generation
python src/main.py \
    --episodes 100 \
    --enable-plotting \
    --plot-interval 10 \
    --verbose
```

## ✅ Verification

The visualization system has been validated through:
- ✅ **Plot generation**: All 20+ plot types render correctly
- ✅ **Data pipeline demo**: Educational flow visualization works
- ✅ **Real-time integration**: Plots update during training without errors
- ✅ **Multiple formats**: PNG, SVG, and HTML output supported

---

**Priority**: Medium  
**Effort**: Medium (already completed)  
**Impact**: High - Essential for training analysis and debugging workflows
